dvc/repo/experiments/executor/base.py

```diff
@@ -283,6 +284,10 @@ def save(
         include_untracked.extend(
             dvc.index._plot_sources  # pylint: disable=protected-access
         )
+        # dvc repro automatically stages dvc.lock. Running redundant `git add`
+        # on it causes an error when exiting the detached head context.
+        if LOCK_FILE in dvc.scm.untracked_files():
+            include_untracked.append(LOCK_FILE)
```

Comment:
You probably need to add every `dvc.lock` file. 
```suggestion
        for file in dvc.scm.untracked_files():
            if os.path.basename(file) == LOCK_FILE:
                include_untracked.append(file)
```

Reply:
I thought about it but we never really defined what would be the expected behavior for `exp save` so I tried to "mimic" `exp run`.

By default  `exp run` only executes the `dvc.yaml` (and tracks the `dvc.lock`) from the cwd  :

```console
git init
dvc init
dvc stage add -n foo echo foo
mkdir subdir
cd subdir 
dvc stage add -n bar echo bar
cd ..
git add .
git commit -m "init"
```

```console
$ dvc exp run
Running stage 'foo':                                                  
> echo foo
foo
Generating lock file 'dvc.lock'                                       
Updating lock file 'dvc.lock'

To track the changes with git, run:

	git add dvc.yaml dvc.lock

To enable auto staging, run:

	dvc config core.autostage true
                                                                      
Ran experiment(s): nervy-cons
Experiment results have been applied to your workspace.

To promote an experiment to a Git branch run:

	dvc exp branch <exp> <branch>
```

Reply:
> That's not true. dvc exp run runs all stages defined in dvc.yaml and all their dependencies that may be defined elsewhere.

Check the script, `dvc exp run` is not executing `subdir/dvc.yaml`

Reply:
Try this script:
```bash
cd "$(mktemp -d)"
git init
dvc init
dvc stage add -n foo -o foo "echo foo > foo"
mkdir subdir
cd subdir
dvc stage add -n bar -d ../foo "cp ../foo bar"
git add -A && git commit -m "init"
dvc exp run
```

Reply:
> Try this script:

Not gonna lie. 
I am not sure if that is even what I would personally want as the default behavior for `dvc exp run` 😅 

---

Is there an easy API call to say "give the lockfiles dependent on this `dvc.yaml`"?

Anyhow, perhaps I am overthinking, and adding any untracked `dvc.lock` is a good default.
Any thoughts @dberenbaum ?

Reply:
> Try this script:

I honestly don't know if I would prefer, as the default behavior for that `dvc exp run`, to fail and say that dependencies are missing.

Reply:
> Anyhow, perhaps I am overthinking, and adding any untracked `dvc.lock` is a good default.
> Any thoughts @dberenbaum ?

That sounds reasonable to me, at least within a given `.dvc` root (saving them across the whole `.git` repo may be problematic for monorepos). Even if the `dvc.lock` hasn't been touched, I'm not sure it's worth granularly saving parts of the project. For `dvc exp run`, will we only save the `dvc.lock` files that have been touched during the run?

Reply:
> For dvc exp run, will we only save the dvc.lock files that have been touched during the run?

If they are untracked (which I think is what we are discussing), yes. The internal `dvc repro` will only stage the lock files of executed stages. 

So, in the example I shared above, a hypothetical `subdir/dvc.lock` will **not** be included in the commit created by `dvc exp run`.